File: microsynth/microsynth/report/find_notes/find_notes.js
// Copyright (c) 2023, Microsynth, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Find Notes"] = {
	"filters": [
		{
            "fieldname": "contact",
            "label": __("Contact"),
            "fieldtype": "Link",
            "options": "Contact"
        },
		{
            "fieldname": "first_name",
            "label": __("First Name"),
            "fieldtype": "Data"
        },
		{
            "fieldname": "last_name",
            "label": __("Last Name"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "customer_name",
            "label": __("Customer Name"),
            "fieldtype": "Data"
        },
		{
            "fieldname": "sales_manager",
            "label": __("Sales Manager"),
            "fieldtype": "Data"
        },
		{
            "fieldname": "territory",
            "label": __("Territory"),
            "fieldtype": "Link",
            "options": "Territory"
        },
        {
            "fieldname": "country",
            "label": __("Country"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "city",
            "label": __("City"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "pincode",
            "label": __("Postal Code"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "street",
            "label": __("Street"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "institute_key",
            "label": __("Institute Key"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "group_leader",
            "label": __("Group Leader"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "from_date",
            "label": __("From date"),
            "fieldtype": "Date"
        },
        {
            "fieldname": "to_date",
            "label": __("To date"),
            "fieldtype": "Date"
        }
	],
    "onload": (report) => {
        report.page.add_inner_button( __("Create PDF"), function() {
            create_pdf(report.get_values());
        });

        if (!locals.double_click_handler) {
            locals.double_click_handler = true;
            // add event listener for double clicks to move up
            cur_page.container.addEventListener("dblclick", function(event) {
                let row = event.delegatedTarget.getAttribute("data-row-index");
                let column = event.delegatedTarget.getAttribute("data-col-index");
                if (parseInt(column) === 16) {
                    let note_id = frappe.query_report.data[row].note_id;
                    let date = frappe.query_report.data[row].date;
                    let contact = frappe.query_report.data[row].contact;
                    let contact_name = frappe.query_report.data[row].first_name + ' ' + frappe.query_report.data[row].last_name;
                    let notes = frappe.query_report.data[row].notes;
                    edit_cell(note_id, date, contact, contact_name, notes);
                }
            });
        }
    }
};


function create_pdf(filters) {
    frappe.call({
        'method': "microsynth.microsynth.report.find_notes.find_notes.create_pdf",
        'args': {
            'filters': filters
        },
        'freeze': true,
        'freeze_message': __("Creating PDF ..."),
        'callback': function (response) {
            // var doc = response.message;
            // frappe.model.sync(doc);
            // frappe.set_route("Form", doc.doctype, doc.name);
            window.location.href = response.message;
        }
    });
}


function edit_cell(note_id, date, contact, contact_name, value) {
    var d = new frappe.ui.Dialog({
        'fields': [
            {'fieldname': 'note_id', 'fieldtype': 'Link', 'options': "Contact Note", 'label': __('Contact Note'), 'read_only': 1, 'default': note_id},
            {'fieldname': 'date', 'fieldtype': 'Date', 'label': __('Date'), 'read_only': 1, 'default': date},
            {'fieldname': 'contact', 'fieldtype': 'Link', 'options': "Contact", 'label': __('Contact'), 'read_only': 1, 'default': contact},
            {'fieldname': 'contact_name', 'fieldtype': 'Data', 'label': __('Contact Name'), 'read_only': 1, 'default': contact_name},
            {'fieldname': 'notes', 'fieldtype': 'Text Editor', 'label': __('Notes'), 'reqd': 1, 'default': value}
        ],
        'primary_action': function(){
            d.hide();
            var values = d.get_values();
            frappe.call({
                'method': "microsynth.microsynth.report.find_notes.find_notes.set_notes",
                'args':{
                    'note_id': values.note_id,
                    'notes': values.notes
                },
                'callback': function(r)
                {
                    frappe.query_report.refresh();
                }
            });
        },
        'primary_action_label': __('Save'),
        'title': __('Edit Notes')
    });
    d.show();
}

File: microsynth/microsynth/report/find_notes/find_notes.py
# Copyright (c) 2023, Microsynth, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _


def get_columns(filters):
    return [
        {"label": _("Note ID"), "fieldname": "name", "fieldtype": "Link", "options": "Contact Note", "width": 100 },
        {"label": _("Date"), "fieldname": "date", "fieldtype": "Date", "width": 75 },
        {"label": _("Contact"), "fieldname": "contact_person", "fieldtype": "Link", "options": "Contact", "width": 65 },
        {"label": _("First Name"), "fieldname": "first_name", "fieldtype": "Data", "width": 80 },
        {"label": _("Last Name"), "fieldname": "last_name", "fieldtype": "Data", "width": 100 },
        {"label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 130 },
        {"label": _("Customer ID"), "fieldname": "customer_id", "fieldtype": "Link", "options": "Customer", "width": 90 },
        {"label": _("Sales Manager"), "fieldname": "sales_manager", "fieldtype": "Data", "options": "User", "width": 100 },
        {"label": _("Territory"), "fieldname": "territory", "fieldtype": "Link", "options": "Territory", "width": 135 },
        {"label": _("Country"), "fieldname": "country", "fieldtype": "Data", "width": 75 },
        {"label": _("City"), "fieldname": "city", "fieldtype": "Data", "width": 100 },
        #{"label": _("Institute"), "fieldname": "institute", "fieldtype": "Data", "width": 175 },
        {"label": _("Institute Key"), "fieldname": "institute_key", "fieldtype": "Data", "width": 100 },
        #{"label": _("Department"), "fieldname": "department", "fieldtype": "Data", "width": 125 },
        {"label": _("Group Leader"), "fieldname": "group_leader", "fieldtype": "Data", "width": 100 },
        {"label": _("Creator"), "fieldname": "owner", "fieldtype": "Data", "options": "User", "width": 100 },
        {"label": _("Note Type"), "fieldname": "contact_note_type", "fieldtype": "Data", "width": 80 },
        {"label": _("Notes"), "fieldname": "notes", "fieldtype": "Data", "options": "Notes", "width": 200 },
    ]


def get_notes(sql_conditions, indent):
    query = f"""
            SELECT
                {indent} AS `indent`,
                `tabContact Note`.`name`,
                `tabContact Note`.`date`,
                `tabContact Note`.`contact_person`,
                `tabContact Note`.`first_name`,
                `tabContact Note`.`last_name`,
                `tabCustomer`.`customer_name`,
                `tabCustomer`.`name` AS `customer_id`,
                `tabCustomer`.`account_manager` AS `sales_manager`,
                `tabCustomer`.`territory`,
                `tabAddress`.`country`,
                `tabAddress`.`city`,
                `tabContact`.`institute`,
                `tabContact`.`institute_key`,
                `tabContact`.`department`,
                `tabContact`.`group_leader`,
                `tabContact Note`.`owner`,
                `tabContact Note`.`contact_note_type`,
                `tabContact Note`.`notes`
            FROM `tabContact Note`
            LEFT JOIN `tabContact` ON `tabContact`.`name` = `tabContact Note`.`contact_person`
            LEFT JOIN `tabDynamic Link` AS `tDLA` ON `tDLA`.`parent` = `tabContact`.`name` 
                                              AND `tDLA`.`parenttype`  = "Contact" 
                                              AND `tDLA`.`link_doctype` = "Customer"
            LEFT JOIN `tabCustomer` ON `tabCustomer`.`name` = `tDLA`.`link_name`
            LEFT JOIN `tabAddress` ON `tabAddress`.`name` = `tabContact`.`address`
            WHERE TRUE
                {sql_conditions}
            ORDER BY `tabContact Note`.`date` DESC
        """
    return frappe.db.sql(query, as_dict=True)


def get_data(filters):
    """
    Get raw Contact Notes records for find notes report.
    """
    filter_conditions = ''

    if filters:
        if filters.get('contact'):
            filter_conditions += f"AND `tabContact Note`.`contact_person` = '{filters.get('contact')}'"
        if filters.get('first_name'):
            filter_conditions += f"AND `tabContact Note`.`first_name` LIKE '{filters.get('first_name')}'"
        if filters.get('last_name'):
            filter_conditions += f"AND `tabContact Note`.`last_name` LIKE '{filters.get('last_name')}'"
        if filters.get('customer_name'):
            filter_conditions += f"AND `tabCustomer`.`customer_name` LIKE '%{filters.get('customer_name')}%'"
        if filters.get('sales_manager'):
            filter_conditions += f"AND `tabCustomer`.`account_manager` LIKE '%{filters.get('sales_manager')}%'"
        if filters.get('territory'):
            filter_conditions += f"AND `tabCustomer`.`territory` = '{filters.get('territory')}'"
        if filters.get('country'):
            filter_conditions += f"AND `tabAddress`.`country` LIKE '%{filters.get('country')}%'"
        if filters.get('city'):
            filter_conditions += f"AND `tabAddress`.`city` LIKE '%{filters.get('city')}%'"
        if filters.get('pincode'):
            filter_conditions += f"AND `tabAddress`.`pincode` = '{filters.get('pincode')}'"
        if filters.get('street'):
            filter_conditions += f"AND `tabAddress`.`address_line1` LIKE '%{filters.get('street')}%'"
        #if filters.get('institute'):
        #    filter_conditions += f"AND `tabContact`.`institute` LIKE '%{filters.get('institute')}%'"
        if filters.get('institute_key'):
            filter_conditions += f"AND `tabContact`.`institute_key` LIKE '%{filters.get('institute_key')}%'"
        #if filters.get('department'):
        #    filter_conditions += f"AND `tabContact`.`department` LIKE '%{filters.get('department')}%'"
        if filters.get('group_leader'):
            filter_conditions += f"AND `tabContact`.`group_leader` LIKE '%{filters.get('group_leader')}%'"
        if filters.get('from_date'):
            filter_conditions += f"AND `tabContact Note`.`date` >= DATE('{filters.get('from_date')}')"
        if filters.get('to_date'):
            filter_conditions += f"AND `tabContact Note`.`date` <= DATE('{filters.get('to_date')}')"

    raw_notes = get_notes(filter_conditions, 0)

    if not filters.get('no_previous_notes') or filters.get('no_previous_notes') < 1:
        return raw_notes
    
    enriched = []
    for note in raw_notes:
        enriched.append(note)
        sql_conditions = f" AND `tabContact Note`.`name` != '{note['name']}'"
        sql_conditions += f" AND `tabContact Note`.`contact_person` = '{note['contact_person']}'"
        previous_notes = get_notes(sql_conditions, 1)        

        for i, previous_note in enumerate(previous_notes):
            if i >= filters.get('no_previous_notes'):
                break
            enriched.append(previous_note)
    
    return enriched


def execute(filters=None):
    columns, data = get_columns(filters), get_data(filters)
    return columns, data


@frappe.whitelist()
def create_pdf(filters):
    """
    Get Contact Notes matching the given filters, create and append their print formats and return the URL to the merged PDF.
 
    bench execute microsynth.microsynth.report.find_notes.find_notes.create_pdf --kwargs "{'filters': {'sales_manager': 'atila.durmus@microsynth.seqlab.de', 'from_date':'2024-06-01', 'to_date':'2024-06-30'}}"
    """
    import io
    import json
    from datetime import datetime
    from collections import OrderedDict
    from frappe.utils.pdf import get_pdf
    from PyPDF2 import PdfFileMerger, PdfFileReader  # JPe: I've checked that PyPDF is available on the server. It was also used in microsynth/report/fiscal_representation_export/fiscal_representation_export.py

    if type(filters) == str:
        filters = json.loads(filters)
    raw_data = get_data(filters)
    enriched_data = OrderedDict()
    merger = PdfFileMerger()

    for i, note in enumerate(raw_data):
    #     if not note['contact_person'] in enriched_data:
    #         enriched_data[note['contact_person']] = note
    #     else:
    #         note_content = enriched_data[note['contact_person']]['notes']
    #         url = f'<a href="https://erp.microsynth.local/desk#Form/Contact%20Note/{note["name"]}"><u>{note["name"]}</u></a>'
    #         note_content += f"<br><b>Contact Note {url} from {note['date'].strftime('%d.%m.%Y')}:</b><p>{note['notes']}</p>"
    #         enriched_data[note['contact_person']]['notes'] = note_content

    # for i, note in enumerate(list(enriched_data.values())):
        # css = frappe.get_value('Print Format', 'Contact Note', 'css')
        # raw_html = frappe.get_value('Print Format', 'Contact Note', 'html')
        # # create html
        # css_html = f"<style>{css}</style>{raw_html}"
        # #frappe.throw(f"{note=}")
        # frappe.log_error(css_html, "css_html")
        # frappe.log_error(note, "note")
        # rendered_html = frappe.render_template(  # TODO: TypeError: 'NoneType' object is not callable
        #     css_html,
        #     {
        #         'doc': note,
        #         'idx': (i+1),
        #         'customer_id': note['customer_id']
        #     }
        # )
        # # need to load the styles and tags
        # content = frappe.render_template(
        #     'microsynth/templates/pages/print.html',
        #     {'html': rendered_html}
        # )
        # options = {
        #     'disable-smart-shrinking': ''
        # }
        # pdf = get_pdf(content, options)
        pdf = frappe.get_print(
                    doctype="Contact Note",
                    name=note['name'],
                    print_format="Contact Note",
                    as_pdf=True
                )
        merger.append(PdfFileReader(io.BytesIO(pdf)))
    # https://gist.github.com/vijaywm/d6b7d1a54274838cd25acd1e3dd31740#pypdf2-pdffilemerger-to-merge-pdfs
    out = io.BytesIO()
    merger.write(out)
    file = frappe.get_doc(
        {
            "doctype": "File",
            "file_name": f"customer_contact_report_{datetime.now()}.pdf",
            "is_private": 1,
            "content": out.getvalue(),
        })
    file.save()
    merger.close()
    #return file.name
    return file.file_url


@frappe.whitelist()
def set_notes(note_id, notes):
    contact_note = frappe.get_doc("Contact Note", note_id)
    contact_note.notes = notes
    contact_note.save()
    frappe.db.commit()
